README.md
# otp_util

`otp_util` is a dart package to generate and verify one-time passwords that were used to implement 2FA and MFA authentication method in web applications and other login-required systems.

The package was implement based on [RFC4226](https://tools.ietf.org/html/rfc4226) (HOTP: An HMAC-Based One-Time Password Algorithm) and [RFC6238](https://tools.ietf.org/html/rfc6238) (TOTP: Time-Based One-Time Password Algorithm).

## Feature

* Create and verify a HOTP object
* Create and verify a TOTP object
* Generate a `otpauth url` with the b32 encoded string
* Support for OTP tokens encrypted with SHA1, SHA256, SHA384 and SHA512

### Installation

#### Pubspec

Add `otp_util` as a dependency in your `pubspec.yaml` file.

```yaml
dependencies:
  otp_util: ^1.0.2
```

### Example

#### Time-based OTPs

```dart
import 'package:dart_otp/dart_otp.dart';

void main() {

  /// default initialization for intervals of 30 seconds and 6 digit tokens
  TOTP totp = TOTP(secret: "BASE32ENCODEDSECRET");

  /// initialization for custom interval and digit values
  TOTP totp = TOTP(secret: "BASE32ENCODEDSECRET", interval: 60, digits: 8);

  totp.now(); /// => 745872

  /// verify for the current time
  totp.verify(otp: '745872'); /// => true

  /// verify after 30s
  totp.verify(otp: '745872'); /// => false
}
```

#### Counter-based OTPs

```dart
import 'package:dart_otp/dart_otp.dart';

void main() {

  /// default initialization for intervals of 30 seconds and 6 digit tokens
  HOTP hotp = HOTP(secret: "BASE32ENCODEDSECRET");

  /// initialization for custom counter and digit value
  HOTP hotp = HOTP(secret: "BASE32ENCODEDSECRET", counter: 50, digits: 8);

  hotp.at(counter: 0); /// => 145089
  hotp.at(counter: 1); /// => 899133
  hotp.at(counter: 2438); /// => 146071

  /// verify with a counter
  hotp.verify(otp: '397732', counter: 2438); /// => true
  hotp.verify(otp: '397732', counter: 2438); /// => false
}
```

### Release notes

See [CHANGELOG.md](./CHANGELOG.md).


CHANGELOG.md
## 1.0.2

* Perfect details.


test/otp_util_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:otp_util/otp_util.dart';
import 'package:otp_util/src/utils/otp_util.dart';

void main() {
  test('adds one to input values', () {
    expect(OTPUtil.otpTypeValue(), 'totp');
    expect(OTPUtil.otpTypeValue(type: OTPType.HOTP), 'hotp');
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
  });
}


lib/src/otp.dart
///
/// @module   : OTP module to generate the password
/// @author   : dushiling (dsl1197467923@gmail.com)
///

import 'dart:math';
import 'package:base32/base32.dart';
import 'package:otp_util/src/utils/algorithm_util.dart';
import 'package:otp_util/src/utils/generic_util.dart';
import 'package:otp_util/src/utils/otp_util.dart';

import 'components/otp_algorithm.dart';
import 'components/otp_type.dart';

abstract class OTP {
  /// The length of the one-time password, between 6 and 8.
  int? digits;

  /// The Base32 secret key used to generate the one-time password.
  String? secret;

  /// The crypto algorithm used on HMAC encoding.
  OTPAlgorithm? algorithm;

  /// The type of the token.
  OTPType get type;

  /// To access custom properties when generating the url.
  Map<String, dynamic> get extraUrlProperties;

  ///
  /// This constructor will create an OTP instance.
  ///
  /// All parameters are mandatory however [digits] and
  /// [algorithm] have a default value, so can be ignored.
  /// Will throw an exception if the line above isn't satisfied.
  ///
  OTP(
      {String? secret,
      int digits = 6,
      OTPAlgorithm algorithm = OTPAlgorithm.SHA1})
      : assert(secret != null),
        assert(digits >= 6 && digits <= 8) {
    this.secret = secret!;
    this.digits = digits;
    this.algorithm = algorithm;
  }

  ///
  /// When class HOTP or TOTP pass the input params to this
  /// function, it will generate the OTP object with params,
  /// the params may be counter or time.
  ///
  /// All parameters are mandatory however [algorithm] have
  /// a default value, so can be ignored.
  ///
  String generateOTP({int? input, OTPAlgorithm algorithm = OTPAlgorithm.SHA1}) {
    /// base32 decode the secret
    var hmacKey = base32.decode(this.secret!);

    /// initial the HMAC-SHA1 object
    var hmacSha =
        AlgorithmUtil.createHmacFor(algorithm: algorithm, key: hmacKey);

    /// get hmac answer
    var hmac = hmacSha.convert(Util.intToBytelist(input: input!)).bytes;

    /// calculate the init offset
    int offset = hmac[hmac.length - 1] & 0xf;

    /// calculate the code
    int code = ((hmac[offset] & 0x7f) << 24 |
        (hmac[offset + 1] & 0xff) << 16 |
        (hmac[offset + 2] & 0xff) << 8 |
        (hmac[offset + 3] & 0xff));

    /// get the initial string code
    var strCode = (code % pow(10, this.digits!)).toString();
    strCode = strCode.padLeft(this.digits!, '0');

    return strCode;
  }

  ///
  /// Generate a url with OTP instance.
  ///
  /// Use [issuer] and [account] parameters to specify the token information.
  /// All the remaining OTP fields will be exported.
  ///
  String generateUrl({String? issuer, String? account}) {
    final _secret = this.secret;
    final _type = OTPUtil.otpTypeValue(type: type);
    final _account = Uri.encodeComponent(account ?? '');
    final _issuer = Uri.encodeQueryComponent(issuer ?? '');

    final _algorithm = AlgorithmUtil.rawValue(algorithm: algorithm);
    final _extra = extraUrlProperties
        .map((key, value) => MapEntry(key, "$key=$value"))
        .values
        .join('&');

    return 'otpauth://$_type/$_account?secret=$_secret&issuer=$_issuer&digits=$digits&algorithm=$_algorithm&$_extra';
  }
}


lib/src/hotp.dart
///
/// @module   : HOTP module to generate and verify HOTP password
/// @author   : dushiling (dsl1197467923@gmail.com)
///

import 'components/otp_algorithm.dart';
import 'components/otp_type.dart';
import 'otp.dart';

///
/// HOTP class will generate the OTP (One Time Password) object with given counter.
class HOTP extends OTP {
  /// The counter value for one-time password.
  int? counter;

  @override
  OTPType get type => OTPType.HOTP;

  @override
  Map<String, dynamic> get extraUrlProperties => {"counter": counter};

  ///
  /// This constructor will create an HOTP instance.
  ///
  /// All parameters are mandatory however [counter],
  /// [digits] and [algorithm] have a default values, so can be ignored.
  ///
  /// Will throw an exception if the line above isn't satisfied.
  ///
  HOTP(
      {String? secret,
      int counter = 0,
      int digits = 6,
      OTPAlgorithm algorithm = OTPAlgorithm.SHA1})
      : super(secret: secret, digits: digits, algorithm: algorithm) {
    this.counter = counter;
  }

  ///
  /// Generate the HOTP value with the given count
  ///
  /// ```dart
  /// HOTP hotp = HOTP(secret: 'BASE32ENCODEDSECRET');
  /// hotp.at(counter: 0); // => 145089
  /// ```
  ///
  String at({int? counter}) {
    if (counter == null || counter < 0) {
      return '';
    }

    return super.generateOTP(input: counter);
  }

  ///
  /// Verifies the HOTP value passed in against the a given counter.
  ///
  /// All parameters are mandatory.
  ///
  /// ```dart
  /// HOTP hotp = HOTP(secret: 'BASE32ENCODEDSECRET');
  /// hotp.at(counter: 0); // => 145089
  /// // Verify for current time
  /// hotp.verify(otp: 145089, counter: 0); // => true
  /// // Verify after 30s
  /// hotp.verify(otp: 145089, counter: 10); // => false
  /// ```
  ///
  bool verify({String? otp, int? counter}) {
    if (otp == null || counter == null) {
      return false;
    }

    String otpCount = this.at(counter: counter);
    return otp == otpCount;
  }
}


lib/src/totp.dart
///
/// @module   : TOTP module to generate and verify TOTP password
/// @author   : dushiling (dsl1197467923@gmail.com)
///

import 'package:otp_util/src/utils/generic_util.dart';

import 'components/otp_algorithm.dart';
import 'otp.dart';
import 'components/otp_type.dart';

///
/// TOTP class will generate the OTP (One Time Password) object with the current or given time.
class TOTP extends OTP {
  /// Period in which should be generated new tokens.
  int interval = 60;

  @override
  OTPType get type => OTPType.TOTP;

  @override
  Map<String, dynamic> get extraUrlProperties => {"period": interval};

  ///
  /// This constructor will create an TOTP instance.
  ///
  /// All parameters are mandatory however [interval],
  /// [digits] and [algorithm] have a default values, so can be ignored.
  ///
  /// Will throw an exception if the line above isn't satisfied.
  ///
  TOTP(
      {String secret = '',
      int digits = 6,
      int interval = 30,
      OTPAlgorithm algorithm = OTPAlgorithm.SHA1})
      : super(secret: secret, digits: digits, algorithm: algorithm) {
    this.interval = interval;
  }

  ///
  /// Generate the TOTP value with current time.
  ///
  /// ```dart
  /// TOTP totp = TOTP(secret: 'BASE32ENCODEDSECRET');
  /// totp.now(); // => 432143
  /// ```
  ///
  String now() {
    int _formatTime = Util.timeFormat(time: DateTime.now(), interval: interval);
    return super.generateOTP(input: _formatTime);
  }

  ///
  /// Generate the OTP with a custom time.
  ///
  /// All parameters are mandatory.
  ///
  /// ```dart
  /// TOTP totp = TOTP(secret: 'BASE32ENCODEDSECRET');
  /// totp.value(date: DateTime.now()); // => 803013
  /// ```
  ///
  String value({DateTime? date}) {
    if (date == null) {
      return '';
    }

    int _formatTime = Util.timeFormat(time: date, interval: interval);
    return super.generateOTP(input: _formatTime);
  }

  ///
  /// Verifies the TOTP value passed in against the current time.
  ///
  /// All parameters are mandatory.
  ///
  /// ```dart
  /// TOTP totp = TOTP(secret: 'BASE32ENCODEDSECRET');
  /// totp.now(); // => 803013
  /// // Verify for current time
  /// totp.verify(otp: 803013); // => true
  /// // Verify after 30s
  /// totp.verify(otp: 803013); // => false
  /// ```
  ///
  bool verify({String? otp, DateTime? time}) {
    if (otp == null) {
      return false;
    }

    var _time = time ?? DateTime.now();
    var _input = Util.timeFormat(time: _time, interval: interval);

    String otpTime = super.generateOTP(input: _input);
    return otp == otpTime;
  }
}


lib/src/utils/generic_util.dart
abstract class Util {
  ///
  /// format the time string to int
  ///
  /// @param {time}
  /// @type {Date}
  /// @desc the time need to be format
  ///
  /// @param {interval}
  /// @type {Int}
  /// @desc interval means the one-time password's life,
  /// default to be 30.
  ///
  /// @return {Int}
  ///
  static int timeFormat({DateTime? time, int? interval}) {
    final _timeStr = time!.millisecondsSinceEpoch.toString();
    final _formatTime = _timeStr.substring(0, _timeStr.length - 3);

    return int.parse(_formatTime) ~/ interval!;
  }

  ///
  /// transfer the int type to List type
  ///
  /// @param {input}
  /// @type {int}
  /// @desc input param, maybe counter or time
  ///
  /// @return {List}
  ///
  static List<int> intToBytelist({required int input, int padding = 8}) {
    List<int> _result = [];
    var _input = input;
    while (_input != 0) {
      _result.add(_input & 0xff);
      _input >>= padding;
    }
    _result.addAll(List<int>.generate(padding, (_) => 0));
    _result = _result.sublist(0, padding);
    _result = _result.reversed.toList();
    return _result;
  }
}


lib/src/utils/algorithm_util.dart
import 'package:crypto/crypto.dart';
import '../components/otp_algorithm.dart';

abstract class AlgorithmUtil {
  static Hmac createHmacFor(
      {OTPAlgorithm? algorithm = OTPAlgorithm.SHA1, required List<int> key}) {
    switch (algorithm) {
      case OTPAlgorithm.SHA1:
        return Hmac(sha1, key);
      case OTPAlgorithm.SHA256:
        return Hmac(sha256, key);
      case OTPAlgorithm.SHA384:
        return Hmac(sha384, key);
      case OTPAlgorithm.SHA512:
        return Hmac(sha512, key);

      default:
        return Hmac(sha1, key);
    }
  }

  static String rawValue({OTPAlgorithm? algorithm = OTPAlgorithm.SHA1}) {
    switch (algorithm) {
      case OTPAlgorithm.SHA1:
        return 'SHA1';
      case OTPAlgorithm.SHA256:
        return 'SHA256';
      case OTPAlgorithm.SHA384:
        return 'SHA384';
      case OTPAlgorithm.SHA512:
        return 'SHA512';

      default:
        return 'SHA1';
    }
  }
}


example/lib/main.dart
import 'package:otp_util/otp_util.dart';

void main() {
  TOTP totp = TOTP(secret: "BASE32ENCODEDSECRET");
  print(totp.now());
  print(totp.value(date: DateTime.now()));
  print(totp.verify(otp: totp.now(), time: DateTime.now()));

  TOTP totp2 = TOTP(secret: "BASE32ENCODEDSECRET", interval: 60, digits: 8);
  print(totp2.now());
  print(totp2.value(date: DateTime.now()));
  print(totp2.value(date: DateTime.parse('2018-10-10 09:30:36')));
  print(totp2.verify(otp: '80271722', time: DateTime.parse('2018-10-10 09:30:36')));

  print('*******************************');

  HOTP hotp = HOTP(secret: "BASE32ENCODEDSECRET");
  print(hotp.at(counter: 0));
  print(hotp.at(counter: 1));
  print(hotp.at(counter: 2438));
  print(hotp.verify(otp: '397732', counter: 2438));

  HOTP hotp2 = HOTP(secret: "BASE32ENCODEDSECRET", counter: 60, digits: 8);
  print(hotp2.at(counter: 0));
  print(hotp2.at(counter: 1));
  print(hotp2.at(counter: 2438));
  print(hotp2.verify(otp: '75397732', counter: 2438));
}
